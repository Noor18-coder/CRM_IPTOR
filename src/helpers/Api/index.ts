export * from './ApiRequest';
export * from './User';
export * from './CompanyInfo';
export * from './CurrencyInfo';
export * from './StagesInfo';
export * from './OpportunityType';
export * from './Attributes';
