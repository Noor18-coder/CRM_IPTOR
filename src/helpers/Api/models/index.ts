export * from './UserItem';
export * from './CompanyInfo';
export * from './ApiRequestControls';
export * from './Opportunity';
export * from './OpportunityDetails';
export * from './Users';
export * from './StageInfo';
export * from './Contacts';
export * from './Product';
export * from './Customer';
export * from './OpportunityTypes';
export * from './CustomerDetails';
export * from './AddOpportunity';
export * from './Items';
export * from './OpportunityAttributes';
export * from './Currency';
export * from './Attributes';
export * from './AddCustomer';
export * from './Countries';
export * from './ForeCastInfo';
