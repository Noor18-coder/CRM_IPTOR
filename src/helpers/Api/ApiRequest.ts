/**
 * Api Request Object
 */
import { ApiRequestControls } from './models';

export class ApiRequest<T> {
  IptorAPI: string;

  id: string;

  constructor(public method: string, public params?: T, public control?: ApiRequestControls) {
    this.IptorAPI = process.env.REACT_APP_IPTOR_API_VERSION ?? '';
    this.id = process.env.REACT_APP_IPTOR_API_ID ?? '';
  }
}
