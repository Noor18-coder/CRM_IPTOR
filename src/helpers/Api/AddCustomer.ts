import axios, { AxiosResponse } from 'axios';
import { get } from 'lodash';
import { AddBusinessPartnerDefaultParams, AddBusinessPartnerResponse, SaveUserDefinedFieldParam } from './models';
import { ApiRequest } from './ApiRequest';

export default class AddCustomerApi {
  /** API Method */
  private static apiMethod = 'crmBusinessPartner.add';

  private static updateApiMethod = 'crmBusinessPartner.update';

  private static saveAttributeValueMethod = 'mopAttribute.add';

  private static saveAttributeValueMethodFile = 'SRONAM';

  /**
   * Helper function to fetch Business Partner Info
   * API Method: crmBusinessPartner.add
   * */
  static async add(businessPartner: AddBusinessPartnerDefaultParams): Promise<AddBusinessPartnerResponse> {
    const requestData = new ApiRequest<AddBusinessPartnerDefaultParams>(this.apiMethod, businessPartner);
    const response = await axios.post<AxiosResponse>('/api/service', requestData);
    return get(response, 'data');
  }

  static async update(businessPartner: AddBusinessPartnerDefaultParams): Promise<AddBusinessPartnerResponse> {
    const requestData = new ApiRequest<AddBusinessPartnerDefaultParams>(this.updateApiMethod, businessPartner);
    const response = await axios.post<AxiosResponse>('/api/service', requestData);
    return get(response, 'data');
  }

  static async addAttributes(businessPartnerId: string, attributeType: string, attributeValue: string | number): Promise<AddBusinessPartnerResponse> {
    const params: SaveUserDefinedFieldParam = {
      attributeType,
      parentFile: this.saveAttributeValueMethodFile,
      parentId: businessPartnerId,
      attributeValue,
    };
    const requestData = new ApiRequest<SaveUserDefinedFieldParam>(this.saveAttributeValueMethod, params);
    const response = await axios.post<AxiosResponse>('/api/service', requestData);
    return get(response, 'data');
  }
}
