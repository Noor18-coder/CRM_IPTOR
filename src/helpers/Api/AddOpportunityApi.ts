import axios, { AxiosResponse } from 'axios';
import { AddOpportunityDefaultParams, AddOpportunityResponse, SaveUserDefinedFieldParam, AddItemToOpportunityParams, AddItemToOpportunityResponse, AddCustomerContactParams, AddCustomerContactRequestParams } from './models';
import { ApiRequest } from './ApiRequest';
import { get } from 'lodash';

export default class AddOpportunityApi {
  /** API Method */
  private static apiMethod: string = 'mopOpportunity.add';
  private static saveAttributeValueMethod: string = 'mopAttribute.add';
  private static saveAttributeValueMethodFile: string = 'SROMOPH';
  private static saveItemMethod: string = 'mopItem.add';
  private static saveContactFileName: string = 'SROMOPCH';
  private static saveContactMethod: string = 'mopContact.add';

  /**
   * Helper function to fetch Business Partner Info
   * API Method: mopOpportunity.get
   * @param freeTextSearch search term
   * 
   */
  static async add(opportunity: AddOpportunityDefaultParams): Promise<AddOpportunityResponse> {
    const requestData = new ApiRequest<AddOpportunityDefaultParams>(this.apiMethod, opportunity);
    const response = await axios.post<AxiosResponse>('/api/service', requestData);
    return get(response, 'data');
  }

  /**
   * Helper function to fetch Business Partner Info
   * API Method: mopOpportunity.get
   * @param freeTextSearch search term
   * 
   */
  static async addAttributes(opportunityId: string, attributeType: string, attributeValue: string | number): Promise<AddOpportunityResponse> {
    const params: SaveUserDefinedFieldParam = {
      attributeType: attributeType,
      parentFile: this.saveAttributeValueMethodFile,
      parentId: opportunityId,
      attributeValue: attributeValue
    }
    const requestData = new ApiRequest<SaveUserDefinedFieldParam>(this.saveAttributeValueMethod, params);
    const response = await axios.post<AxiosResponse>('/api/service', requestData);
    return get(response, 'data');
  }

  /**
   * Helper function to fetch Business Partner Info
   * API Method: mopOpportunity.get
   * @param freeTextSearch search term
   * 
   */
  static async addItem(opportunityId: string, item: string, quantity: number, unit:string): Promise<AddItemToOpportunityResponse> {
    const params: AddItemToOpportunityParams = {
      parentFile: this.saveAttributeValueMethodFile,
      parentId: opportunityId,
      item: item,
      quantity: quantity,
      unit:unit
    }
    const requestData = new ApiRequest<AddItemToOpportunityParams>(this.saveItemMethod, params);
    const response = await axios.post<AddItemToOpportunityResponse>('/api/service', requestData);
    return get(response, 'data');
  }

  /**
   * Helper function to fetch Business Partner Info
   * API Method: mopOpportunity.get
   * @param freeTextSearch search term
   * 
   */
   static async addContacts(params:AddCustomerContactParams): Promise<AxiosResponse> {
    const request:AddCustomerContactRequestParams  = {
      contactParentFile:this.saveAttributeValueMethodFile,
      contactParentId:params.contactPArentId,
      contactPerson:params.contactPerson,
      phone:params.phone,
      mobile:params.mobile,
      contactDC:'',
      whatsApp:'',
      linkedin:'',
      fax:'',
      email:params.email
    }

    const requestData = new ApiRequest<AddCustomerContactRequestParams>(this.saveContactMethod, request);
    const response = await axios.post<AxiosResponse>('/api/service', requestData);
    console.log(response);
    return get(response, 'data');
  }
}
