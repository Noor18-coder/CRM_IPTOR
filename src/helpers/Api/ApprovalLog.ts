import axios, { AxiosResponse } from 'axios';
import { get } from 'lodash';
import { ApprovalLogAddRequestParams, ApprovalLogAddResponse, DeleteApprovalLogsParams } from './models';
import { ApiRequest } from './ApiRequest';

export class ApprovalLog {
  /** API Method */
  private static apiMethod = 'crmOpportunityApprovalLog.add';

  private static deleteApprovalLogs = 'crmOpportunityApprovalLog.delete';

  /**
   * Helper function to fetch sales order list
   * @param params user id
   * @returns information about user
   */
  static async addApprovalLog(params: ApprovalLogAddRequestParams): Promise<ApprovalLogAddResponse> {
    const requestData = new ApiRequest<ApprovalLogAddRequestParams>(this.apiMethod, params);
    const response = await axios.post<ApprovalLogAddResponse>('/api/service', requestData);
    return get<AxiosResponse<ApprovalLogAddResponse>, 'data'>(response, 'data');
  }

  static async deleteOpportunityApprovalLogs(params: DeleteApprovalLogsParams): Promise<ApprovalLogAddResponse> {
    const requestData = new ApiRequest<DeleteApprovalLogsParams>(this.deleteApprovalLogs, params);
    const response = await axios.post<ApprovalLogAddResponse>('/api/service', requestData);
    return get(response, 'data.data.items', {});
  }
}
