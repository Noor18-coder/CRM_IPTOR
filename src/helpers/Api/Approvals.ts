import axios, { AxiosResponse } from 'axios';
import { get } from 'lodash';
import * as apiModels from './models';
import { ApiRequest } from './ApiRequest';

export class ApprovalInfo {
  /** API Method */
  private static allApprovalsMethod = 'crmOpportunityApprovalLogs.get';

  private static allNotesMethod = 'mopNotes.get';

  private static notesParentFile = 'SROMOPAL';

  static async getApprovalLogs(opportunityId: string): Promise<apiModels.ApprovalLogsDefault[]> {
    const requestData = new ApiRequest<apiModels.ApprovalLogsParams>(this.allApprovalsMethod, { opportunityId });
    const response = await axios.post<apiModels.ApprovalLogsResponse>('/api/service', requestData);
    return get<AxiosResponse<apiModels.ApprovalLogsResponse>, 'data', 'data', 'items', apiModels.ApprovalLogsDefault[]>(
      response,
      ['data', 'data', 'items'],
      []
    );
  }

  static async getAllNotes(parentId: string): Promise<apiModels.AllNotesDefault> {
    const requestData = new ApiRequest<apiModels.AllNotesParams>(this.allNotesMethod, { parentFile: this.notesParentFile, parentId });
    const response = await axios.post<apiModels.AllNotesResponse>('/api/service', requestData);
    const defaultData = {} as apiModels.AllNotesDefault;
    return get<AxiosResponse<apiModels.AllNotesResponse>, 'data', 'data', 'items', apiModels.AllNotesDefault>(
      response,
      ['data', 'data', 'items'],
      defaultData
    );
  }
}
