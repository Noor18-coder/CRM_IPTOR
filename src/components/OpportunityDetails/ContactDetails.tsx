import React from 'react';
import { Dispatch } from 'redux';
import { useSelector, useDispatch } from 'react-redux';
import { Card, Image, Accordion } from 'react-bootstrap';
import { DeleteCustomerContactParams, OpportunityContact } from '../../helpers/Api/models';
import OpportunityDetailsApi from '../../helpers/Api/OpportunityDetailsApi';
import ImageConfig from '../../config/ImageConfig';
import * as models from '../../helpers/Api/models';
import { AppState } from '../../store';
import { Attributes } from '../../helpers/Api';

import { deleteContactFromOpportunity, openOpportunityForm, getOpportunityContacts } from '../../store/OpportunityDetails/Actions';

export interface ContactProps {
  data: OpportunityContact;
}

interface Props {
  opportunityId: string;
}

export const ContactAccordian: React.FC<Props> = ({ opportunityId }) => {
  const [activeClass, setActiveClass] = React.useState('');
  const state: AppState = useSelector((appState: AppState) => appState);
  const dispatch: Dispatch<any> = useDispatch();
  const { contacts } = state.opportuntyDetails;
  const toggleAccordion = () => {
    setActiveClass(activeClass === '' ? 'active' : '');
  };

  const openAddContactForm = () => {
    document.body.classList.add('body-scroll-hidden');
    dispatch(openOpportunityForm({ open: true, groupName: 'add_contact', action: 'edit' }));
  };

  React.useEffect(() => {
    dispatch(getOpportunityContacts(opportunityId));
  }, []);

  return (
    <Accordion defaultActiveKey="0">
      <Card className="add-details">
        <Accordion.Toggle className={activeClass} onClick={toggleAccordion} as={Card.Link} eventKey="1">
          Contacts
          {state.opportuntyDetails.editOportunity.allowEdit ? (
            <Image src={ImageConfig.ADD_BTN} alt="Add" title="Add" onClick={openAddContactForm} />
          ) : null}
        </Accordion.Toggle>
        <Accordion.Collapse eventKey="1">
          <div className="container-fluid">
            <div className="row">
              {contacts.length ? (
                contacts.map((obj: OpportunityContact) => {
                  return <ContactCards data={obj} />;
                })
              ) : (
                <div className="no-data-txt"> No Contacts Found </div>
              )}
            </div>
          </div>
        </Accordion.Collapse>
      </Card>
    </Accordion>
  );
};

export const ContactCards: React.FC<React.PropsWithChildren<ContactProps>> = ({ data }) => {
  const state: AppState = useSelector((appState: AppState) => appState);
  const dispatch: Dispatch<any> = useDispatch();
  const [contactDetails, setContactDetails] = React.useState<models.OpportunityDetailsGroupItem[]>([]);
  const [opportunityRole, setOpportunityRole] = React.useState<string | undefined>('');
  React.useEffect(() => {
    loadContactAttributes();
  }, []);

  const loadContactAttributes = async () => {
    const response = await OpportunityDetailsApi.getContactDetails(data.contactId);
    setContactDetails(response);
    const addressObj = response.find((obj: any) => obj.attributeType === 'ADDRESS');
    // eslint-disable-next-line no-param-reassign
    data.visitingAddress = addressObj?.attributeValue;
    const roleObj = response.find((obj) => obj.attributeType === 'ROLE');
    setOpportunityRole(roleObj?.attributeValue);
  };

  const removeContact = async () => {
    const params: DeleteCustomerContactParams = {
      contactId: data.contactId,
      contactParentId: data.contactParentId,
      contactParentFile: data.contactParentFile,
    };
    dispatch(deleteContactFromOpportunity(params));
  };

  const changeRole = async (e: React.ChangeEvent<HTMLSelectElement>) => {
    const newValue = e.currentTarget.value;

    if (opportunityRole) {
      const roleObj = contactDetails.find((obj) => obj.attributeType === 'ROLE');
      const valueId = roleObj?.valueId ? roleObj.valueId : '';
      const params: models.SaveAttributeFieldParam = {
        attributeType: 'ROLE',
        attributeValue: newValue,
        valueId,
      };
      await Attributes.updateAttribute(params);
      setOpportunityRole(newValue);
    } else {
      await Attributes.addAttributes('SROMOPCH', data.contactId, 'ROLE', newValue);
      setOpportunityRole(newValue);
    }
  };

  return (
    <div className="col-md-4">
      <div className="card mb-4">
        <div className="card-body">
          <div className="name-add-sec">
            <p className="person-name-desig">
              <span className="name">{data.contactPerson}</span>
              <span className="designation">Associate - Sales</span>
            </p>
            <p className="mailid">{data.email ? data.email : '--'}</p>
            <p className="contact-num">{data.phone ? data.phone : '--'}</p>
            <p className="contact-address">{data.visitingAddress ? data.visitingAddress : '--'}</p>
          </div>
          {state.opportuntyDetails.editOportunity.allowEdit ? (
            <button type="button" className="address-card-close link-anchor-button" onClick={removeContact}>
              <Image className="card-delete" height="20" src={ImageConfig.CLOSE_BTN} />
            </button>
          ) : null}

          <div className="select-opportunity-sec">
            {state.opportuntyDetails.editOportunity.allowEdit ? (
              <div className="form-group">
                <label htmlFor="contact-role">Select Opportunity Role</label>
                <select className="form-control opportunity-dd" id="contact-role" value={opportunityRole} multiple={false} onChange={changeRole}>
                  {state.enviornmentConfigs.opportunityContactRoles.map((obj: models.DropDownValue) => {
                    return obj.fieldDescription ? (
                      <option value={obj.valueField}>
                        {obj.valueField} - {obj.fieldDescription}
                      </option>
                    ) : (
                      <option value={obj.valueField}>{obj.valueField}</option>
                    );
                  })}
                </select>
              </div>
            ) : (
              <p className="contact-address">{opportunityRole || '--'}</p>
            )}
          </div>
        </div>
      </div>
    </div>
  );
};
