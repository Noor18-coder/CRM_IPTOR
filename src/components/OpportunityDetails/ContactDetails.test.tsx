import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { ContactAccordian, ContactCards } from './ContactDetails';
import StoreMock from '../../mocks/Store.mock';
import { OpportunityContact } from '../../helpers/Api/models';

const store = StoreMock.createInitialStore();
const data = {} as OpportunityContact;

describe('[OpportunityDetails] ContactAccordian', () => {
  it('should renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <ContactAccordian opportunityId="theId" />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('[OpportunityDetails] ContactCards', () => {
  it('should renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <ContactCards data={data} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
