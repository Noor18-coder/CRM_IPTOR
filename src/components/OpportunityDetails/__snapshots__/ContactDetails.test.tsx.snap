// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[OpportunityDetails] ContactAccordian should renders correctly 1`] = `
<div
  className="accordion"
>
  <div
    className="add-details card"
  >
    <a
      className="card-link"
      onClick={[Function]}
    >
      Contacts
    </a>
    <div
      aria-expanded={null}
      className="collapse"
    >
      <div
        className="container-fluid"
      >
        <div
          className="row"
        >
          <div
            className="no-data-txt"
          >
            No contacts.
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`[OpportunityDetails] ContactCards should renders correctly 1`] = `
<div
  className="col-md-4"
>
  <div
    className="card mb-4"
  >
    <div
      className="card-body"
    >
      <div
        className="name-add-sec"
      >
        <p
          className="person-name-desig"
        >
          <span
            className="name"
          />
          <span
            className="designation"
          />
        </p>
        <p
          className="mailid"
        >
          --
        </p>
        <p
          className="contact-num"
        >
          --
        </p>
        <p
          className="contact-address"
        >
          --
        </p>
      </div>
      <div
        className="select-opportunity-sec"
      >
        <p
          className="contact-address"
        >
          --
        </p>
      </div>
    </div>
  </div>
</div>
`;
